File: Trie/Python/trie.py
'''
Author: Tswatery
Date: 2023-11-08 23:05:14
'''
class TrieNode:
    def __init__(self):
        self.children = [ None for i in range(26)] # 26个儿子
        self.is_end = False
        self.exists = 0
    def insert(self, root, s: str) -> None:
        mv = root
        for c in s:
            idx = ord(c) - ord('a')
            if mv.children[idx] == None:
                mv.children[idx] = TrieNode()
            mv = mv.children[idx]
        mv.is_end = True
        mv.exists = 1
    def find(self, root, s: str) -> int:
        mv = root
        for c in s:
            idx = ord(c) - ord('a')
            if mv.children[idx] == None:
                return -1
            mv = mv.children[idx]
        if mv.exists == 1:
            mv.exists += 1
            return 1
        return mv.exists

File: Trie/C++/trie.cpp
/*
 * @Author: Tswatery
 * @Date: 2023-11-08 22:29:36
 */
#include "trie.h"
#include <iostream>
#include <cstring>
#include <algorithm>

bool TrieNode::checkNull(char c){
    return children[c - 'a'] == nullptr;
}

void TrieNode::setNode(char c){
    children[c - 'a'] = new TrieNode();
}

TrieNode* TrieNode::getNode(char c){
    return children[c - 'a'];
}

void TrieNode::setEnd(){
    is_end = true;
}

bool TrieNode::getEnd(){
    return is_end;
}

void TrieNode::insert(TrieNode* root, std::string& s) {
    TrieNode* mv = root;
    for(int i = 0; i < (int)s.size(); ++ i) {
        if(! mv->children[s[i] - 'a']) {
            mv->children[s[i] - 'a'] = new TrieNode();
        }
        mv = mv->children[s[i] - 'a'];
    }
    mv->is_end = true;
}

bool TrieNode::find(TrieNode* root, std::string& s) {
    TrieNode* mv = root;
    for(int i = 0; i < (int)s.size(); ++ i) {
        if(! mv->children[s[i] - 'a']) return false; // 不存在的话就返回false
        mv = mv->children[s[i] - 'a'];
    }
    return mv->is_end;
}